File: incident_api/schemas/user_activity.py
"""
Pydantic schemas for the User Activity and Cross-Reference feature.
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime

from . import IncidentInDB # Importar el schema de Incidentes

# --- Individual Activity Schemas ---

class ActivityLogItem(BaseModel):
    """
    Represents a single item in the user's activity timeline (from IncidentLog).
    """
    log_id: int
    incident_id: int
    action: str
    comments: Optional[str] = None
    timestamp: datetime

    class Config:
        from_attributes = True

class HistoryLogItem(BaseModel):
    """
    Represents a single change in the user's history (from IncidentHistory).
    """
    history_id: int = Field(..., alias='id')
    incident_id: int
    field_changed: str
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    timestamp: datetime

    class Config:
        from_attributes = True

# --- Metrics Schema ---

class UserActivityMetrics(BaseModel):
    """
    Defines the calculated metrics and basic info for a user's activity.
    """
    # User Info
    user_id: int
    full_name: str
    email: str
    role: str
    is_active: bool

    # Calculated Metrics
    total_incidents_created: int
    total_incidents_resolved: int
    total_incidents_assigned: int
    last_login: Optional[datetime] = None
    top_incident_types: List[str]
    average_resolution_time_hours: Optional[float] = None
    incidents_by_status: Dict[str, int]
    login_frequency_per_week: Optional[float] = None
    total_comments_made: int
    total_files_uploaded: int

# --- Main Response Schema ---

class IncidentRelationship(IncidentInDB):
    """Extends IncidentInDB to include the relationship type for the graph."""
    relationship_type: str

class UserCrossReferenceResponse(BaseModel):
    """
    The main response model for the GET /users/{user_id}/cross-reference endpoint.
    It consolidates all cross-referenced information for a user.
    """
    metrics: UserActivityMetrics
    incidents: List[IncidentRelationship]
    activity_logs: List[ActivityLogItem]
    history_logs: List[HistoryLogItem]

File: react_frontend/src/types/user_activity.ts
/**
 * Types for the User Cross-Reference feature.
 */

// A base incident type, can be expanded as needed
export interface Incident {
    id: number;
    title: string;
    status: string;
}

export interface ActivityLogItem {
    log_id: number;
    incident_id: number;
    action: string;
    comments?: string;
    timestamp: string; // ISO date string
}

export interface HistoryLogItem {
    history_id: number;
    incident_id: number;
    field_changed: string;
    old_value?: string;
    new_value?: string;
    timestamp: string; // ISO date string
}

export interface UserActivityMetrics {
    user_id: number;
    full_name: string;
    email: string;
    role: string;
    is_active: boolean;
    total_incidents_created: number;
    total_incidents_resolved: number;
    total_incidents_assigned: number;
    last_login?: string; // ISO date string
    top_incident_types: string[];
    average_resolution_time_hours?: number;
    incidents_by_status: Record<string, number>;
    login_frequency_per_week?: number;
    total_comments_made: number;
    total_files_uploaded: number;
}

export interface IncidentRelationship extends Incident {
    relationship_type: string;
}

export interface UserCrossReferenceResponse {
    metrics: UserActivityMetrics;
    incidents: IncidentRelationship[];
    activity_logs: ActivityLogItem[];
    history_logs: HistoryLogItem[];
}
